### ISSUE TITLE
Wifi Plug Output support with Tasmota (new Tasmota Output module)

### ISSUE THREAD

**Issue by smguth** [kizniche/Mycodo#719]
**Is your feature request related to a problem? Please describe.**
 Wired relays and cable management hastles could be alleviated. 
**Describe the solution you'd like**
I would like integration with Mycodo outputs.  
This is a wifi plug that can recieve commands via SSH.   I was considering using gpio wired relays but this seems like a perfect solution to avoid all the wires.  I just purchased a unit on amazon, waiting for it to arrive.  https://www.amazon.com/gp/product/B00MWR6ZY4/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1
My thought is I can use output commands with Mycodo and it will work just fine but also want the Mycodo community to know about this smart plug because I think they would like it and come up with ideas for it that I possibly would not
.  
Future enhancement suggestion.   Integration with Mycodo

**Additional context**
Please see this link it tells about the wifi plug further
http://www.anites.com/2015/01/hacking-kankun-smart-wifi-plug.html

**Comment by kizniche** [comment]
I love the wireless outlets. The Command Output that's currently available in Mycodo would be the most appropriate for this. Essentially, a script needs to be developed that turns it on and off, then that script can be executed by the Mycodo Output Controller to control the plug. When you get the plug, we can start developing a working script to test.

**Comment by kizniche** [comment]
It looks like the bash script from your second link should work well:

```bash
#!/bin/sh
echo "Content-Type: text/plain"
echo "Cache-Control: no-cache, must-revalidate"
echo "Expires: Sat, 26 Jul 1997 05:00:00 GMT"
echo

RELAY_CTRL=/sys/class/leds/tp-link:blue:relay/brightness

case "$QUERY_STRING" in
 state) 
  case "`cat $RELAY_CTRL`" in
   0) echo "OFF"
   ;;
   1) echo "ON"
   ;;
  esac
 ;;
 on) 
  echo 1 > $RELAY_CTRL
  echo OK
 ;;
 off) 
  echo 0 > $RELAY_CTRL
  echo OK
 ;;
esac
```

**Comment by smguth** [comment]
Yeah I was wondering if it was worth bothing to add a check to ensure that
the relay state changed after the control was sent.   If i did bother with
it then do i send the control again and check the state again.  After so
many failed attempts trigger an alert or email.  Probably just start with
having it send the command and work from there though.  I have a feeling it
will always work as long as it is in range

**Comment by SAM26K** [comment]
That AC plug looks awesome as it has ssh options.

**Comment by SAM26K** [comment]
@smguth do you have a sales link to the US version?  Unable to locate so  far. Thanks.

**Comment by smguth** [comment]
@SAM23K I only found it on amazon with the link from above https://www.amazon.com/gp/product/B00MWR6ZY4/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1.   With the holidays I am still waiting for it to arrive too.  There was mention of finding it on  AliExpress for about $20 in the other link but I didn't go down that route.  I am planning on verifying that everything works smoothly when the one I ordered from Amazon comes in then probably try to source them cheaper as $30 per plug is a bit more than I ideally want to spend.  If anyone does find a better source for these please let us know!

**Comment by kizniche** [comment]
There seem to be a lot of wifi plugs on the market, I'm sure there are many that have already been hacked or have an open API of sorts out-of-the-box.

https://www.amazon.com/s?k=wifi+plug

**Comment by kizniche** [comment]
Where were these things a few years ago!? Some interesting finds:

https://www.amazon.com/AHRISE-Protector-6-Outlet-Compatible-Assistant/dp/B07VV2MGMC
https://www.amazon.com/POWRUI-4-Outlet-Weatherproof-Assistant-Network，Black/dp/B07KK62GB7
https://www.amazon.com/Compatible-Wireless-Smartphone-Waterproof-certified/dp/B07S915JZ7
https://www.amazon.com/Kasa-Smart-Power-Strip-TP-Link/dp/B07G95FFN3
https://www.amazon.com/IMILLET-Protector-Outlets-Compatible-Assistant/dp/B074CJYPZV

**Comment by kizniche** [comment]
I found my holy grail of power strips, with the rare use of solid state relays, however it looks like this product has been discontinued: https://github.com/arendst/Tasmota/wiki/LeFun-SK2-Smart-Power-Strip-(Model-ZLD-44USA-W)

**Comment by SAM26K** [comment]
@smguth How long ago did you order it? Says 10 days for delivery on Amazon.

**Comment by SAM26K** [comment]
A lot of the wireless  plugs can be hacked by burning Tasmota firmware but often it's very difficult to connect to the data ports.  The Kankun is the first one I've heard of that has an open interface and reflashing isn't necessary.

**Comment by kizniche** [comment]
I bought the LeFun power strip with hope I'll be able to put Tasmota on it. The SSRs are really appealing for my applications.

**Comment by SAM26K** [comment]
That one looks good and on the Tasmota list of deviices.

**Comment by smguth** [comment]
wow, now I want a smart power strip lol.  Looks like they can monitor power usage too!

**Comment by smguth** [comment]
Wow I hadn't heard of Tasmota before but now I need to learn about it.  Do some of these smart plugs monitor power usage too?

**Comment by smguth** [comment]
https://blakadder.github.io/templates/kmc-4.html  found this guy says it a energy monitoring smart plug might order this next after my kankun comes in

**Comment by smguth** [comment]
https://blakadder.github.io/templates/ce_power_stake.html seems like this would be good for humid environments

**Comment by SAM26K** [comment]
If you pick a device for Tasmota flashing google it  fist. I look for detailed instructions with pictures.

**Comment by SAM26K** [comment]
Most of the wireless  AC adapters  use the ESP8266 chip which is the target for Tasmota.  Some devices are easy to connect to and others are very difficult to connect for reflashing.

**Comment by SAM26K** [comment]
BTW the energy monitoring software is stock with many of the device's  original firmware but not sure if it's  included with Tasmota.  You can use Mycodo to set current draw  for a  device and have a calculated value for power logged to the database.

**Comment by smguth** [comment]
Thanks SAM26K I think i'm starting to get a feel for some of this stuff now :D .   I'm still very new to Mycodo so might be off base with my understanding of it.  Trying to get my project off the ground yet but having fun learning as I go.  So if my device is Tasmota compatible does that mean it is compatible with Mycodo or might there still be configuration/integration yet?

**Comment by kizniche** [comment]
If you can get Tasmota on it, which will allow for switching on/off via HTTP or SSH, then Mycodo can be configured to use it as an output.

**Comment by SAM26K** [comment]
@smguth please flollow up when you finally get your  Kankun device.   May work with Mycodo  out of the  box without  Tasmota.

**Comment by smguth** [comment]
Okay I'll keep you updated Amazon says it will arrive the 16-19 so possibly today.

**Comment by smguth** [comment]
Got this guy in over the weekend had a few hurdles to overcome with the set up.  For what ever reason the android app wouldn't work for me so I had to configure it with ssh and get it on my wifi network.  The instructions worked but I had to set my password key to psk2 instead of what their guide said.  Also, to control the relay I had to use
root@koven:~# echo 0 > /sys/devices/platform/leds-gpio/leds/i-konke\:red\:relay/
brightness
root@koven:~# echo 1 > /sys/devices/platform/leds-gpio/leds/i-konke\:red\:relay/
brightness

My model must have been slightly different than the one in the guide

**Comment by smguth** [comment]
my /etc/config/network file looks as follow: 

config interface 'loopback'
	option ifname 'lo'
	option proto 'static'
	option ipaddr '127.0.0.1'
	option netmask '255.0.0.0'

config globals 'globals'
	option ula_prefix 'fd54:f988:6027::/48'

config interface 'lan'
	option ifname 'eth0'
	option force_link '1'
	option type 'bridge'
	option proto 'static'
	option ipaddr '192.168.145.253'
	option netmask '255.255.255.0'
	option ip6assign '60'

config interface 'wan'
	option proto 'dhcp'
	option ifname 'eth1'

config interface 'wwan'
	option proto 'dhcp'

and my /etc/config/wireless:

config wifi-device 'radio0'
	option type 'mac80211'
	option channel '11'
	option hwmode '11ng'
	option path 'platform/ar933x_wmac'
	list ht_capab 'SHORT-GI-20'
	list ht_capab 'SHORT-GI-40'
	list ht_capab 'RX-STBC1'
	list ht_capab 'DSSS_CCK-40'
	option htmode 'HT20'
	option disabled '0'
	option country 'CN'

config wifi-iface
	option device 'radio0'
	option network 'wwan'
	option ssid 'Network Name'
	option encryption 'psk2'
	option mode 'sta'
	option key 'password'

**Comment by smguth** [comment]
![image](https://user-images.githubusercontent.com/11531815/71327542-6e904400-24cf-11ea-9257-53f318cc17bc.png)
Hi guys,  I created a script on the Pi that ssh's into the kankun and runs a script that turns on the relay.  If I run the script from my pi it turns it on and off with both sets of scripts.  When I try to run the script from mycodo it does not work.  I'm wondering if its a user or permissions thing.

**Comment by SAM26K** [comment]
Looks promising.  Can you disable the password requirement?  You are probably right about permissions. Also try using Python code instead of ssh.

**Comment by SAM26K** [comment]
Also is there errors in the daemon log?

**Comment by smguth** [comment]
![image](https://user-images.githubusercontent.com/11531815/71413996-76103400-261a-11ea-9fd6-5754397c51ec.png)
I didn't see any errors in the log.  It looks like mycodo thinks its just executing the commands succesfully from my understanding of it.  I'll look into writing this in python gonna have to do the holiday thing here shortly though.

**Comment by SAM26K** [comment]
Having problems with my Linux controls since  installing the new Raspbian o/s and Mycodo 8.1.x. In the process of converting them to Python.

**Comment by SAM26K** [comment]
Do you know what's causing the GPIO error?

**Comment by smguth** [comment]
I have a Dht11 connected to pin 17 and it's giving me data so I was a little surprised to see the error.  I never put a resistor in series with the the data line now that I think about it so maybe I'll do that and see if it fixes it

**Comment by kizniche** [comment]
I can add the option to execute the commands as the user pi or root. The default user commands are executed as is currently pi.

**Comment by SAM26K** [comment]
That would be nice.

**Comment by SAM26K** [comment]
@smguth any progress? Advise not to update with the vendors firmware.  You may be able to install Tasmota OTA (over the air)) without the need to wire directly into the device.

**Comment by smguth** [comment]
Hey @SAM26K I havnt gotten much further.  I was hoping to try replace my scripts with python code but I've never coded in python before so that's been a hurdle.  I can try getting tasmota on it.  That will protect the manufacturer from locking me out?

**Comment by kizniche** [comment]
> I can add the option to execute the commands as user pi or root.

I had forgotten about this since it's not really related to this issue, but I was going through issues today and found it. Better late than never. I just committed an update that adds the ability to select which user (pi, mycodo, or root) executes Linux Command Outputs.

I also forgot to give an update on the power strip I ordered. The seller cancelled the order and that was the only supplier I could find for that particular version.

**Comment by smguth** [comment]
Well I got this working.  I reinstalled my operating system and started over with a clean mycodo install.  There were alot of steps but I followed the notes I had in this issue.  I also had to run
`ssh root@wifiplug "tee -a /etc/dropbear/authorized_keys" < ~/.ssh/id_rsa.pub`
From my raspberrypi in order for the pi to send the commands to the kankun smart plug without a password.   Also, I added wifiplug to hosts file so I didn't have to keep using the ip address.   No when I click 
![image](https://user-images.githubusercontent.com/11531815/81973094-3fe34900-95e9-11ea-8020-7a24ac49623d.png)
 On and off buttons the relay in the kankun plug turns on and off

**Comment by kizniche** [comment]
I'm reviving this thread since I just ordered 4 individual wifi outlets that include power monitoring.

https://templates.blakadder.com/efun_SH331W.html
https://www.amazon.com/gp/product/B07DCJ7TDR

Hopefully I won't have a reoccurrence of what happened last time and the seller cancels my order. I'll let you kow how flashing of Tasmota and building the output module go after I receive the product.

**Comment by kizniche** [comment]
All 4 plugs were quick to flash Tasmota and I can now control them via MQTT publish commands from an MQTT server (mosquito) hosted on the same Raspberry Pi Mycodo is installed on. Each plug had the MQTT server set to localhost, as well as the host for the MQTT outputs. The Mycodo output topic was set to "cmnd/tasmota1/Power" with tasmota1 being your device name, and the ON Payload to "on" and OFF payload to "off".

It's a very nice system and I haven't had any issues in the past week or so they've been running. I'm closing this issue since we've figured out connectivity with Tasmota.

**Comment by aTanguay** [comment]
Hi-

Is there any way you can elaborate on setting this up? Very new to MQTT and I am puzzled. I have Tasmoto installed on my plugs, and mosquito server installed running on the same Pi as Mycodo, but I can't seem to get these to talk to each other.

Here's my MQTT setup in the plug,

<img width="335" alt="firefox_LxAfRMycr5" src="https://user-images.githubusercontent.com/10527352/150424130-98108fbf-c1de-46dc-a772-9ad7b79a27f8.png">

and here's my server output config..

<img width="905" alt="firefox_6yxSqTLrtK" src="https://user-images.githubusercontent.com/10527352/150424242-a86904bb-1b84-4cd0-a083-d7a908158762.png">

From here, I'm just trying to use the On/Off buttons on the output manually. But I get nothing. 

Sorry for the newbie question, but this is definitely some new ground for me.

**Comment by aTanguay** [comment]
I was close. Apparently, I had not installed my MQTT broker properly, so it wasn't able to accept outside connections. Reinstalled it and now it's good. For others trying this, I followed this to get MQTT going...

https://randomnerdtutorials.com/how-to-install-mosquitto-broker-on-raspberry-pi/

From there I got the info from my info screen on the switch...

<img width="432" alt="image" src="https://user-images.githubusercontent.com/10527352/150459132-6a8e0383-59bf-45f6-8b73-7eaca95d20fd.png">

Its configured this way...

<img width="432" alt="image" src="https://user-images.githubusercontent.com/10527352/150459232-669c8533-635a-43c2-93bb-cc0d65d9b6cf.png">

Very exciting, works great, over the air.
